node/sub_peers.js

```diff
@@ -0,0 +1,129 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var inherits = require('util').inherits;
+
+var TChannelPeersBase = require('./peers_base.js');
+
+function TChannelSubPeers(channel, options) {
+    if (!(this instanceof TChannelSubPeers)) {
+        return new TChannelSubPeers(channel, options);
+    }
+    var self = this;
+    TChannelPeersBase.call(self, channel, options);
+
+    self.peerScoreThreshold = self.options.peerScoreThreshold || 0;
+}
+
+inherits(TChannelSubPeers, TChannelPeersBase);
+
+TChannelSubPeers.prototype.close = function close(callback) {
+    var self = this;
+
+    var peers = self.values();
+    TChannelPeersBase.prototype.close.call(self, peers, callback);
+};
+
+TChannelSubPeers.prototype.add = function add(hostPort, options) {
+    /*eslint max-statements: [2, 25]*/
+    var self = this;
+
+    var peer = self._map[hostPort];
+    if (peer) {
+        return peer;
+    }
+
+    var topChannel = self.channel.topChannel;
+
+    if (hostPort === topChannel.hostPort) {
+        return topChannel.peers.getSelfPeer();
+    }
+
+    peer = topChannel.peers.add(hostPort, options);
+
+    self._map[hostPort] = peer;
+    self._keys.push(hostPort);
+
+    return peer;
+};
+
+TChannelSubPeers.prototype.clear = function clear() {
+    var self = this;
+
+    self._map = Object.create(null);
+    self._keys = [];
+};
+
+TChannelSubPeers.prototype._delete = function _del(peer) {
+    var self = this;
+
+    delete self._map[peer.hostPort];
+    var index = self._keys.indexOf(peer.hostPort);
+    popout(self._keys, index);
+};
+
+TChannelSubPeers.prototype.choosePeer = function choosePeer(req) {
+    /*eslint complexity: [2, 15]*/
+    var self = this;
+
+    var hosts = self._keys;
+    if (!hosts || !hosts.length) {
+        return null;
+    }
+
+    var threshold = self.peerScoreThreshold;
+
+    var selectedPeer = null;
+    var selectedScore = 0;
+    for (var i = 0; i < hosts.length; i++) {
+        var hostPort = hosts[i];
+        var peer = self._map[hostPort];
+        if (!req || !req.triedRemoteAddrs[hostPort]) {
+            var score = peer.state.shouldRequest(req);
+            var want = score > threshold &&
+                       (selectedPeer === null || score > selectedScore);
+            if (want) {
+                selectedPeer = peer;
+                selectedScore = score;
+            }
+        }
+    }
+    return selectedPeer;
+};
+
+function popout(array, i) {
+    if (!array.length) {
+        return;
+    }
+
+    if (array.length === 1) {
+        array.pop();
+        return;
+    }
+
+    var j = array.length - 1;
+    var tmp = array[i];
+    array[i] = array[j];
+    array[j] = tmp;
+}
```

Comment:
Copy paste makes copy pasta (and makes less hot for v8 optimizer). Factor it out.